File: ios/BFExplorer/Plaoc/IPCCore/core/NativeMicroModule.swift
//
//  NativeMicroModule.swift
//  BFExplorer
//
//  Created by ui03 on 2023/2/25.
//

import UIKit
import Vapor

typealias RouterHandler<T> = (_ request: Request, _ ipc: Ipc?) -> T?

class NativeMicroModule: MicroModule {
    
    var reqidRouting: [/* req_id */Int:String/* route */] = [:]
    
    var apiRouting: [/* route */String:RouterHandler<Any>] = [:]
    
    func createNativeMicro() {
        
        _ = connectAdapterManager.append { fromMM, toMM, reason in
            if let toMM = toMM as? NativeMicroModule {
                let channel = NativeMessageChannel<IpcMessage, IpcMessage>()
                let toNativeIpc = NativeIpc(port: channel.port1, remote: fromMM, role: .SERVER)
                let fromNativeIpc = NativeIpc(port: channel.port2, remote: toMM, role: .CLIENT)
                fromMM.beConnect(ipc: fromNativeIpc, reason: reason)
                toMM.beConnect(ipc: toNativeIpc, reason: reason)
                return ConnectResult(ipcForFromMM: fromNativeIpc, ipcForToMM: toNativeIpc)
            } else {
                return nil
            }
        }
    }
    
    init(mmid: String) {
        super.init()
        self.mmid = mmid
        
        _ = onConnect { clientIpc, req in
            _ = clientIpc.onRequest { request,ipc in
                guard let url = URL(string: request.urlString) else { return }
                self.reqidRouting[request.req_id] = url.host! + url.path
                let ipcRequest = request.toRequest()
                print("fetch", "NMM/Handler", ipcRequest.url)
                let res = self.defineHandler(request: ipcRequest, ipc: clientIpc)
                let ipcResMessage = IpcResponse.fromResponse(req_id: request.req_id, response: res, ipc: clientIpc)!.ipcResMessage
                clientIpc.postMessage(message: ipcResMessage)
            }
        }
    }
    
    // 对路由处理方法包裹一层，用于http路由
    func defineHandler(request: Request, ipc: Ipc? = nil) -> Response {
        let routeHandler = self.apiRouting[request.route!.path.string]!
        
        let result = routeHandler(request, ipc)
        
        if let result = result as? Response {
            return result
        } else if let result = result as? Codable {
            let content = ChangeTools.tempAnyToString(value: result)
            return Response(status: .ok, headers: .init([("Content-Type", "application/json")]), body: .init(string: content!))
        } else {
            return Response(status: .internalServerError, body: .init(string: """
                <p>\(request.url)</p>
                <pre>Unknow Error</pre>
            """.trimmingCharacters(in: .whitespacesAndNewlines)))
        }
    }
/*
    internal func defineHandler( request: Request, handler: (Request) -> Any?) -> Response {
        
        var response: Response?
        let result = handler(request)
        if let res = result as? Response {
            response = res
        } else {
            var headers = HTTPHeaders()
            headers.add(name: "Content-Type", value: "application/json")
            
            let status = HTTPResponseStatus(statusCode: 200)
            let content = ChangeTools.tempAnyToString(value: result)
            if content != nil {
                let body = Response.Body.init(string: content!)
                
                response = Response(status: status, headers: headers, body: body)
            } else {
                let status = HTTPResponseStatus(statusCode: 500)
                let whitespace = NSCharacterSet.whitespacesAndNewlines
                let content = """
                            <p>${request.uri}</p>
                            <pre>${ex.message ?: "Unknown Error"}</pre>
                            """.trimmingCharacters(in: whitespace)
                let body = Response.Body.init(string: content)
                response = Response(status: status, headers: HTTPHeaders(), body: body)
            }
        }
        return response!
    }
    
    internal func defineHandler(req: Request, handler: (Request, Ipc) -> Any?) -> Response {
        
        return defineHandler(request: req) { request in
            return handler(request, Ipc())
        }
    }*/
}

File: android/MicroService/src/main/java/org/dweb_browser/microservice/core/NativeMicroModule.kt
package org.dweb_browser.microservice.core

import kotlinx.serialization.json.JsonElement
import org.dweb_browser.helper.printDebug
import org.dweb_browser.helper.runBlockingCatching
import org.dweb_browser.microservice.help.gson
import org.dweb_browser.microservice.help.jsonBody
import org.dweb_browser.microservice.help.types.MMID
import org.dweb_browser.microservice.help.types.MicroModuleManifest
import org.dweb_browser.microservice.ipc.Ipc
import org.dweb_browser.microservice.ipc.NativeIpc
import org.dweb_browser.microservice.ipc.NativeMessageChannel
import org.dweb_browser.microservice.ipc.helper.IPC_ROLE
import org.dweb_browser.microservice.ipc.helper.IpcMessage
import org.dweb_browser.microservice.ipc.helper.IpcResponse
import org.http4k.core.Filter
import org.http4k.core.MemoryBody
import org.http4k.core.Request
import org.http4k.core.RequestContexts
import org.http4k.core.Response
import org.http4k.core.Status
import org.http4k.core.then
import org.http4k.core.with
import org.http4k.filter.ServerFilters
import org.http4k.lens.RequestContextKey
import org.http4k.lens.RequestContextLens
import org.http4k.routing.RoutingHttpHandler
import java.io.InputStream

fun debugNMM(tag: String, msg: Any = "", err: Throwable? = null) = printDebug("DNS", tag, msg, err)

abstract class NativeMicroModule(manifest: MicroModuleManifest) : MicroModule(manifest) {
  constructor(mmid: MMID, name: String) : this(
    MicroModuleManifest().apply {
      this.mmid = mmid
      this.name = name
    }
  )

  companion object {
    init {
      connectAdapterManager.append { fromMM, toMM, reason ->
        if (toMM is NativeMicroModule) {
          debugNMM("NMM/connectAdapter", "fromMM: ${fromMM.mmid} => toMM: ${toMM.mmid}")
          val channel = NativeMessageChannel<IpcMessage, IpcMessage>();
          val toNativeIpc = NativeIpc(channel.port1, fromMM, IPC_ROLE.SERVER);
          val fromNativeIpc = NativeIpc(channel.port2, toMM, IPC_ROLE.CLIENT);
          fromMM.beConnect(fromNativeIpc, reason) // 通知发起连接者作为Client
          toMM.beConnect(toNativeIpc, reason) // 通知接收者作为Server
          return@append ConnectResult(fromNativeIpc, toNativeIpc) // 返回发起者的ipc
        } else null
      }
    }
  }

  var apiRouting: RoutingHttpHandler? = null


  private val requestContexts = RequestContexts()
  private val requestContextKey_ipc = RequestContextKey.required<Ipc>(requestContexts)
  private val ipcApiFilter = ServerFilters.InitialiseRequestContext(requestContexts)

  /**
   * 实现一整套简易的路由响应规则
   */
  init {
    onConnect { (clientIpc) ->
      clientIpc.onRequest { (ipcRequest) ->
        val routes = apiRouting ?: return@onRequest;
        val routesWithContext = routes.withFilter(ipcApiFilter.then(Filter { next ->
          { next(it.with(requestContextKey_ipc of clientIpc)) }
        }));
        debugNMM("NMM/Handler", ipcRequest.url)
        val request = ipcRequest.toRequest()
        val response = routesWithContext(request)
        clientIpc.postMessage(
          IpcResponse.fromResponse(
            ipcRequest.req_id, response, clientIpc
          )
        )
      }
    }
  }

  class ResponseRegistry {
    companion object {
      private val regMap = mutableMapOf<Class<*>, (item: Any) -> Response>()

      fun <T : Any> registryResponse(type: Class<T>, handler: (item: T) -> Response) {
        regMap[type] = handler as (item: Any) -> Response
      }

      init {
        registryResponse(ByteArray::class.java) {
          Response(Status.OK).body(MemoryBody(it))
        }
        registryResponse(InputStream::class.java) {
          Response(Status.OK).body(it)
        }
      }

      fun <T : Any> registryJsonAble(type: Class<T>, handler: (item: T) -> Any) {
        registryResponse(type) {
          asJson(handler(it))
        }
      }

      fun handle(result: Any): Response {
        val javaClass = result.javaClass
        return when (val handler = regMap[javaClass]) {
          null -> {
            var superJavaClass = javaClass.superclass
            while (superJavaClass != null) {
              // 尝试寻找继承关系
              when (val handler = regMap[superJavaClass]) {
                null -> superJavaClass = superJavaClass.superclass
                else -> return handler(result)
              }
            }
            // 否则默认当成JSON来返回
            return asJson(result)
          }
          // 如果有注册处理函数，那么交给处理函数进行处理
          else -> handler(result)
        }
      }

      fun asJson(result: Any) =
        Response(Status.OK).body(gson.toJson(result)).header("Content-Type", "application/json")

    }
  }

  data class HandlerContext(val request: Request, private val getIpc: RequestContextLens<Ipc>) {
    val ipc get() = getIpc(request)
    fun throwException(
      code: Status = Status.INTERNAL_SERVER_ERROR,
      message: String = code.description,
      cause: Throwable? = null
    ): Nothing = throw ResponseException(code, message, cause)
  }

  protected fun defineEmptyResponse(
    beforeResponse: (Response.() -> Response)? = null,
    handler: suspend HandlerContext.(request: Request) -> Unit,
  ) = wrapHandler(beforeResponse) {
    HandlerContext(it, requestContextKey_ipc).handler(it)
    Response(Status.NO_CONTENT)
  }

  protected fun defineStringResponse(
    beforeResponse: (Response.() -> Response)? = null,
    handler: suspend HandlerContext.(request: Request) -> String
  ) = wrapHandler(beforeResponse) { request ->
    Response(Status.OK).body(HandlerContext(request, requestContextKey_ipc).handler(request))
  }

  protected fun defineBooleanResponse(
    beforeResponse: (Response.() -> Response)? = null,
    handler: suspend HandlerContext.(request: Request) -> Boolean
  ) = wrapHandler(beforeResponse) {
    Response(Status.OK).jsonBody(HandlerContext(it, requestContextKey_ipc).handler(it))
  }

  protected fun defineJsonResponse(
    beforeResponse: (Response.() -> Response)? = null,
    handler: suspend HandlerContext.(request: Request) -> JsonElement
  ) = wrapHandler(beforeResponse) {
    Response(Status.OK).jsonBody(
      HandlerContext(
        it, requestContextKey_ipc
      ).handler(it)
    )
  }

  fun Response.body(body: JsonElement) = jsonBody(body)

  protected fun defineResponse(
    beforeResponse: (Response.() -> Response)? = null,
    handler: suspend HandlerContext.(request: Request) -> Response
  ) = wrapHandler(beforeResponse) {
    HandlerContext(it, requestContextKey_ipc).handler(it)
  }

  protected fun defineByteArrayHandler(
    beforeResponse: (Response.() -> Response)? = null,
    handler: suspend HandlerContext.(request: Request) -> ByteArray
  ) = wrapHandler(beforeResponse) {
    Response(Status.OK).body(MemoryBody(HandlerContext(it, requestContextKey_ipc).handler(it)))
  }

  protected fun defineInputStreamHandler(
    beforeResponse: (Response.() -> Response)? = null,
    handler: suspend HandlerContext.(request: Request) -> InputStream
  ) = wrapHandler(beforeResponse) {
    Response(Status.OK).body(HandlerContext(it, requestContextKey_ipc).handler(it))
  }

  private fun wrapHandler(
    beforeResponse: (Response.() -> Response)? = null,
    handler: suspend (request: Request) -> Response?,
  ) = { request: Request ->
    runBlockingCatching {
      handler(request)?.let { response ->
        if (beforeResponse != null) {
          response.beforeResponse()
        } else response
      } ?: Response(Status.NOT_IMPLEMENTED)
    }.getOrElse { ex ->
      debugNMM("NMM/Error", request.uri, ex)
      Response(Status.INTERNAL_SERVER_ERROR).body(
        """
          <p>${request.uri}</p>
          <pre>${ex.message ?: "Unknown Error"}</pre>
        """.trimIndent()
      )
    }
  }


  protected fun defineHandler(handler: suspend (request: Request) -> Any?) = { request: Request ->
    runBlockingCatching {
      when (val result = handler(request)) {
        null, Unit -> {
          Response(Status.OK)
        }

        is Response -> result
        is ByteArray -> Response(Status.OK).body(MemoryBody(result))
        is InputStream -> Response(Status.OK).body(result)
        else -> {
          // 如果有注册处理函数，那么交给处理函数进行处理
          ResponseRegistry.handle(result)
        }
      }
    }.getOrElse { ex ->
      debugNMM("NMM/Error", request.uri, ex)
      when (ex) {
        is ResponseException -> Response(ex.code)
        else -> Response(Status.INTERNAL_SERVER_ERROR)
      }.body(
        """
          <p>${request.uri}</p>
          <pre>${ex.message ?: "Unknown Error"}</pre>
        """.trimIndent()
      )
    }
  }

  protected fun defineHandler(handler: suspend (request: Request, ipc: Ipc) -> Any?) =
    defineHandler { request ->
      handler(request, requestContextKey_ipc(request))
    }

  class ResponseException(
    val code: Status = Status.INTERNAL_SERVER_ERROR,
    message: String = code.description,
    cause: Throwable? = null
  ) : Exception(message, cause)
}
